edb/lang/ir/staeval.py

```diff
@@ -0,0 +1,263 @@
+    elif op == ast.ops.LE:
+        value = left_val <= right_val
+    else:
+        raise UnsupportedExpressionError(
+            f'unexpected operator: {op}',
+            context=binop.context)
+
+    return irast.BooleanConstant(
+        value='true' if value else 'false',
+        type=schema.get('std::bool')
+    )
+
+
+def evaluate_arithmetic_binop(
+        binop: irast.BinOp,
+        schema: s_schema.Schema) -> irast.BaseConstant:
+
+    op = binop.op
+    left = evaluate(binop.left, schema=schema)
+    left_val = const_to_python(left, schema=schema)
+    right = evaluate(binop.right, schema=schema)
+    right_val = const_to_python(right, schema=schema)
+
+    real_t = schema.get('std::anyreal')
+    int_t = schema.get('std::anyint')
+
+    left_type = irutils.infer_type(left, schema)
+    right_type = irutils.infer_type(right, schema)
+
+    if not left_type.issubclass(real_t) or not right_type.issubclass(real_t):
+        return
+
+    result_type = s_scalars.get_op_type(
+        op, left_type, right_type, schema=schema)
+
+    if op == ast.ops.ADD:
+        value = left_val + right_val
+    elif op == ast.ops.SUB:
+        value = left_val - right_val
+    elif op == ast.ops.MUL:
+        value = left_val * right_val
+    elif op == ast.ops.DIV:
+        if left_type.issubclass(int_t) and right_type.issubclass(int_t):
+            value = left_val // right_val
+        else:
+            value = left_val / right_val
+    elif op == ast.ops.POW:
+        value = left_val ** right_val
+    elif op == ast.ops.MOD:
+        value = left_val % right_val
+    else:
+        raise UnsupportedExpressionError(
+            f'unexpected operator: {op}',
+            context=binop.context)
+
+    if result_type.issubclass(int_t):
+        return irast.IntegerConstant(value=str(value), type=result_type)
+    else:
+        return irast.FloatConstant(value=str(value), type=result_type)
+
+
+@functools.singledispatch
+def const_to_python(
+        ir: irast.BaseConstant,
+        schema: s_schema.Schema) -> object:
+    raise NotImplementedError(
+        f'cannot convert {ir.__class__} to a Python value')
+
+
+@const_to_python.register(irast.IntegerConstant)
+def int_const_to_python(
+        ir: irast.IntegerConstant,
+        schema: s_schema.Schema) -> object:
+
+    if ir.type.name == 'std::decimal':
+        return decimal.Decimal(ir.value)
+    else:
+        return int(ir.value)
+
+
+@const_to_python.register(irast.FloatConstant)
+def float_const_to_python(
+        ir: irast.FloatConstant,
+        schema: s_schema.Schema) -> object:
+
+    if ir.type.name == 'std::decimal':
+        return decimal.Decimal(ir.value)
+    else:
+        return float(ir.value)
+
+
+@const_to_python.register(irast.StringConstant)
+def str_const_to_python(
+        ir: irast.StringConstant,
+        schema: s_schema.Schema) -> str:
+
+    rv = qlquote.quote_literal(ir.value)
+    return pyast.literal_eval(f"''{rv}''")
```

Comment:
I'd use something like this instead of literal eval:

```python
import ast
import re


def unescape(st):

    str_re = re.compile(r'''
        (?P<slash> \\\\) |
        (?P<dq> \\") |
        (?P<sq> \\') |
        (?:\\x(?P<x>[0-7][0-9a-fA-F])) |
        (?:\\u(?P<u>[0-9a-fA-F]{4})) |
        (?:\\U(?P<U>[0-9a-fA-F]{8})) |
        (?P<n>\\n) |
        (?P<t>\\t) |
        (?P<r>\\r)
    ''', re.X)

    subs = {
        'slash': '\\',
        'dq': '"',
        'sq': "'",
        't': '\t',
        'n': '\n',
        'r': '\r',
    }

    def cb(m):
        g = m.lastgroup
        try:
            return subs[g]
        except KeyError:
            pass

        return chr(int(m.group(g), 16))

    return str_re.sub(cb, st)


x = r'aaaa a \u0062 \U00001234 >\ < \\ \' \" \n \t ! \r asd \x62'
print(x)
print('UNESC', unescape(x))
print('LITEV', ast.literal_eval(f'"""{x}"""'))

```